### PR TITLE
Add moniker support

### PR DESCRIPTION
**Pull request by efritz**
This PR 

- Adds moniker output to the lsif-go binary
- Adds a new utility to add read go.mod and go.sum to and rewrite monikers for external use
- Adds support for stdout output for lsif-go (this allows us to stream lsif-go output directly to lsif-gomod)
- Refactors the binary arguments to act like lsif-tsc (now that we have two cmds, this was a good opportunity)

### CODE REVIEW

**Comment by unknwon** [comment]
Haven't looked in to the actual implementation, just want to say the package structure was designed to use subcommands within a single binary to avoid producing multiple binaries for easy distribution. That's why the equivalent command to `lsif-tsc` is `lsif-go export` not `lsif-go`, and `lsif-go link` was the design/idea for the package linker (sorry didn't put this into the doc somewhere). With that said, the subcommand approach is not necessary good. cc @chrismwendt

**Comment by felixfbecker** [comment]
I think either the monikers should always be added, or it should be a separate tool (not just a subcommand). The thought behind separate tools is that you may use different package managers, and you have one tool per package manager. I don't know if that applied to Go, or if everyone is already using Go modules?

**Comment by unknwon** [comment]
>The thought behind separate tools is that you may use different package managers, and you have one tool per package manager. I don't know if that applied to Go, or if everyone is already using Go modules?

The current source code parser `go/packages` behaves the exact same way as `go` command which takes care of build environment, it is aware of where to look for dependencies, GOPATH, vendor and/or Go modules (i.e. `GOPATH/pkg/mod`). 

If the user wants to be absolutely explicit about which mechanism for looking up dependencies, we could also have something like `lsif-go link -driver=gomod` or `lsif-go link -driver=vendor`.

**Comment by felixfbecker** [comment]
Aren't there other package managers than gomod (and vendoring)? How do you figure out the version for a vendored module?

**Comment by unknwon** [comment]
>Aren't there other package managers than gomod (and vendoring)? How do you figure out the version for a vendored module?

No matter which vendor/pm tool a project/package uses, ultimately it has to be able to compiled by "go" command. And if you're talking about the dependency version info extraction (e.g. `Gopkg.lock`, `vendor.json`), then the `lsif-go link -driver=xx` would still work.

**Comment by efritz** [comment]
> I think either the monikers should always be added, or it should be a separate tool.

The separate tool is the recommendation by Microsoft, which I think we should follow.

**Comment by unknwon** [comment]
Just to be clear, I'm fine with either separate binaries or subcommands.

File: cmd/lsif-go/main.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
+// The program lsif-go is an LSIF exporter for Go.
```
Comment (unknwon) [inline cmd/lsif-go/main.go]: Should we change this "exporter" to "indexer"?

File: cmd/lsif-gomod/main.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package main
```
Comment (unknwon) [inline cmd/lsif-gomod/main.go]: Do we need a line of comment for description of what lsif-gomod is/does?

File: cmd/lsif-gomod/main.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+func realMain() error {
+	var (
+		projectRoot string
+		inFile      string
+		stdin       bool
+		outFile     string
+		stdout      bool
+	)
+
+	app := kingpin.New("lsif-go", "lsif-gomod decorates lsif-go output with gomod monikers.").Version(versionString)
```
Comment (unknwon) [inline cmd/lsif-gomod/main.go]:
```suggestion
	app := kingpin.New("lsif-gomod", "lsif-gomod decorates lsif-go output with gomod monikers.").Version(versionString)
```

File: export/exporter.go [hunk 4]
```diff
@@ -591,3 +590,13 @@ func (e *exporter) emitItemOfReferences(outV string, inVs []string, docID string
 	id := e.nextID()
 	return id, e.emit(protocol.NewItemOfReferences(id, outV, inVs, docID))
 }
+
+func (e *exporter) emitMoniker(kind, sourceID, identifier string) error {
+	monikerID := e.nextID()
+	err := e.emit(protocol.NewMoniker(monikerID, kind, "go", identifier))
```
Comment (unknwon) [inline export/exporter.go]:
```suggestion
	err := e.emit(protocol.NewMoniker(monikerID, kind, protocol.LanguageID, identifier))
```

~~Or if it's always "go", we could reduce one parameter and set this value inside `protocol.NewMoniker`.~~

File: gomod/moniker.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	}
+}
+
+func (d *decorator) decorate(line string) error {
+	moniker := &moniker{}
+	if err := json.Unmarshal([]byte(line), &moniker); err != nil {
+		return fmt.Errorf("unmarshal line: %v", err)
+	}
+
+	if moniker.Type == "vertex" && moniker.Label == "moniker" {
```
Comment (unknwon) [inline gomod/moniker.go]:
```suggestion
	if moniker.Type == protocol.ElementVertex && moniker.Label == protocol.VertexMoniker {
```

File: protocol/protocol.go [hunk 6]
```diff
@@ -357,6 +357,60 @@ func NewHoverResult(id string, contents []MarkedString) *HoverResult {
...
+	Vertex
+	// The name of the package.
+	Name string `json:"name"`
+	// The package manager.
+	Manager string `json:"manager"`
+	// The version of the package.
+	Version string `json:"version"`
+}
+
+// NewPackageInformation returns a new PackageInformation with the given ID< name, manager, and version.
```
Comment (unknwon) [inline protocol/protocol.go]:
```suggestion
// NewPackageInformation returns a new PackageInformation with the given ID, name, manager, and version.
```

File: cmd/lsif-gomod/main.go (hunk 2 above)
Comment (efritz) [reply cmd/lsif-gomod/main.go]: Done.

File: cmd/lsif-go/main.go (hunk 1 above)
Comment (efritz) [reply cmd/lsif-go/main.go]: Will change all of these in another PR.

**Pull request merged**